### PR TITLE
Patch: fix #252

### LINKED ISSUE #252: Foraging DeerHunt. Hunting negative deers.
Hi!

I was messing about with the configs such that the foragingMaxDeerPerHunt == foragingDeerMaxPopulation (if you wish to reproduce this error). I was curious what would happen. 

It seems like agents are allowed to hunt negative deers? 

image
image

### PR DESCRIPTION
**Pull request by JamesTev**
# Summary

Fixed #252 where deer returns where always zero when max deer per hunt was set to max population. Note that this should not usually be the case per the design, but this fix ensures it doesn't break the forage in any case. 

## Test Plan

- Tested with max pop and max per hunt equal. Works as expected.

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

**Comment by JamesTev** [comment]
@E-Mac27 @mpardalos Please have a quick check when you get a moment.

File: internal/common/foraging/deerhunt.go [hunk 1]
```diff
@@ -66,6 +66,10 @@ func (d DeerHunt) getPopulationLinkedProbability(dhConf config.DeerHuntConfig, p
 	thetaCritical := dhConf.ThetaCritical
 	thetaMax := dhConf.ThetaMax
 	p := float64(population) / float64(dhConf.MaxDeerPerHunt)
+
+	if pMax == pCritical { // catch the special case where max deer per hunt = max population. You've got params wrong if you're here though :)
```
Comment (E-Mac27) [inline internal/common/foraging/deerhunt.go:70]: Haha love the comment

Comment (lhl2617) [inline internal/common/foraging/deerhunt.go:70]: let's log this so they know :D

**Review by E-Mac27 — APPROVED** [review]

File: internal/common/foraging/deerhunt.go (hunk 1 above)
Comment (lhl2617) [reply internal/common/foraging/deerhunt.go:70]: (you can pass in the server's logger func into `DeerHunt` and make a logger function for DeerHunt)

The output would ideally be

```
[SERVER] [DEERHUNT] ...
```

Comment (E-Mac27) [reply internal/common/foraging/deerhunt.go:70]: I was thinking the same. Just spam the logs but it also might be too much spam? tbf, its already spammy af so feel free to add more.

File: params.go [hunk 2]
```diff
@@ -51,7 +51,7 @@ var (
 	foragingDeerMaxPerHunt = flag.Uint(
 		"foragingMaxDeerPerHunt",
 		4,
-		"Max possible number of deer on a single hunt (regardless of number of participants).",
+		"Max possible number of deer on a single hunt (regardless of number of participants). ** should be strictly less than max deer population.",
```
Comment (lhl2617) [inline params.go:54]: you should add a check in the `parseConfig` function which now allows you to return parse errors.

**Review by lhl2617 — CHANGES_REQUESTED** [review]
still impressed with foraging

File: internal/common/foraging/deerhunt.go (hunk 1 above)
Comment (lhl2617) [reply internal/common/foraging/deerhunt.go:70]: spammy is fine as long as it is _easily filterable_ ( `| grep [DEERHUNT]`)

File: params.go (hunk 2 above)
Comment (lhl2617) [reply params.go:54]: if you have this, you can remove your above check (i'd still leave it in)

File: internal/common/foraging/deerhunt.go (hunk 1 above)
Comment (JamesTev) [reply internal/common/foraging/deerhunt.go:70]: Okay sure - makes sense!

Comment (lhl2617) [reply internal/common/foraging/deerhunt.go:70]: feel free to do this in a future PR--i'm happy with this as long as you add the extra check in `parseConfig`.

File: params.go (hunk 2 above)
Comment (JamesTev) [reply params.go:54]: Good idea - will do soon

File: internal/common/foraging/deerhunt.go [hunk 3]
```diff
@@ -81,5 +86,5 @@ func (d DeerHunt) getPopulationLinkedProbability(dhConf config.DeerHuntConfig, p
 
 // Logf is a this type's custom logger
 func (d DeerHunt) Logf(format string, a ...interface{}) {
-	log.Printf("[SERVER][DeerHunt]: %v", fmt.Sprintf(format, a...))
+	log.Printf("[SERVER][DEERHUNT]: %v", fmt.Sprintf(format, a...))
```
Comment (lhl2617) [inline internal/common/foraging/deerhunt.go:89]: ah, this is fine, but i was actually thinking that when the server first initialises deerhunt, it passes the pointer to the server's logf function (s.Logf). d.Logf would then call the logger with just `[DEERHUNT]` preprended.

Comment (lhl2617) [reply internal/common/foraging/deerhunt.go:89]: so `NewDeerHunt(s.Logf, ...)`

```
type DeerHunt struct {
     logger func(format string, ...)
}
func NewDeerHunt(logger, ...) {
    return DeerHunt{ logger: logger, ... }
}
func (d *DeerHunt) Logf(format, ...) {
    d.logger.Logf("[DEERHUNT]: ...")
}
```

Comment (lhl2617) [reply internal/common/foraging/deerhunt.go:89]: separate PR encouraged :) you can also show how it should be done for  iigointernal, I think they're struggling....

Comment (JamesTev) [reply internal/common/foraging/deerhunt.go:89]: Okay cool. Will do in a separate PR. Let me know if anything else to add in this one pls :) @lhl2617

Comment (lhl2617) [reply internal/common/foraging/deerhunt.go:89]: it's good let's goooooo

**Review by lhl2617 — APPROVED** [review]

**Pull request merged**